core/extension/gdextension_manager.cpp

```diff
@@ -387,6 +387,33 @@ bool GDExtensionManager::ensure_extensions_loaded(const HashSet<String> &p_exten
 	return needs_restart;
 }
 
+void GDExtensionManager::startup() {
+	for (const KeyValue<String, Ref<GDExtension>> &E : gdextension_map) {
+		const Ref<GDExtension> &extension = E.value;
+		if (extension->startup_callback) {
+			extension->startup_callback();
+		}
+	}
+}
+
+void GDExtensionManager::shutdown() {
```

Comment:
Same for shutdown. If a GDExtension is being unloaded, but _not_ because all of Godot is being shutdown, we still need to run this. This could probably be done with a flag too

We also need to make sure to clear callbacks when reloading a GDExtension

Reply:
I don't have a use case for `startup` or `shutdown` so if that's the behavior that you want I'm happy to implement it.

However, I feel like it may not be a good fit under something called `main_loop_callbacks` if it's also called when the GDExtension is loaded/unloaded.

I feel like this behavior overlaps with the `initialize` and `deinitialize` callbacks from the `GDExtensionInitialization` struct and I'm not sure what benefit they provide over using the existing ones.

Reply:
> I don't have a use case for `startup` or `shutdown`

Sorry, if this is taking you too off course, you could remove those from this PR, and I could make another PR just focused on those? :-)

> I feel like this behavior overlaps with the `initialize` and `deinitialize` callbacks from the `GDExtensionInitialization` struct and I'm not sure what benefit they provide over using the existing ones.

The benefit is that Godot is guaranteed to be fully initialized when they run, which isn't the case for the `initialize` and `deinitialize` callbacks: during Godot's startup and shutdown process, those will get called when Godot is in various in-between states

Reply:
That makes sense. I've implemented the flag you mention, calling the callbacks on loading/unloading the extension, and clearing the callbacks on unloading.

Let me know if this is what you had in mind, I'm not sure if I'm doing it in the right place. Then I'll squash the commits.